nav2_behaviors/plugins/assisted_teleop.cpp

```diff
@@ -0,0 +1,186 @@
+
+namespace nav2_behaviors
+{
+AssistedTeleop::AssistedTeleop()
+: TimedBehavior<AssistedTeleopAction>(),
+  feedback_(std::make_shared<AssistedTeleopAction::Feedback>())
+{}
+
+void AssistedTeleop::onConfigure()
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  // set up parameters
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "projection_time", rclcpp::ParameterValue(1.0));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "input_vel_topic", rclcpp::ParameterValue(std::string("input_vel_topic")));
+
+  nav2_util::declare_parameter_if_not_declared(
+    node,
+    "joystick_topic", rclcpp::ParameterValue(std::string("joystick_topic")));
+
+  node->get_parameter("projection_time", projection_time_);
+  node->get_parameter("input_vel_topic", input_vel_topic_);
+  node->get_parameter("joystick_topic", joystick_topic_);
+
+  vel_sub_ = node->create_subscription<geometry_msgs::msg::Twist>(
+    input_vel_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::inputVelocityCallback,
+      this, std::placeholders::_1));
+
+  joy_sub_ = node->create_subscription<sensor_msgs::msg::Joy>(
+    joystick_topic_, rclcpp::SystemDefaultsQoS(),
+    std::bind(
+      &AssistedTeleop::joyCallback,
+      this, std::placeholders::_1));
+}
+
+Status AssistedTeleop::onRun(const std::shared_ptr<const AssistedTeleopAction::Goal> command)
+{
+  command_time_allowance_ = command->time_allowance;
+  end_time_ = steady_clock_.now() + command_time_allowance_;
+  return Status::SUCCEEDED;
+}
+
+void AssistedTeleop::onActionCompletion()
+{
+  RCLCPP_INFO(logger_, "Action completed");
+  input_twist_ = geometry_msgs::msg::Twist();
+  joy_ = sensor_msgs::msg::Joy();
+}
+
+Status AssistedTeleop::onCycleUpdate()
+{
+  rclcpp::Duration time_remaining = end_time_ - steady_clock_.now();
+  if (time_remaining.seconds() < 0.0 && command_time_allowance_.seconds() > 0.0) {
+    stopRobot();
+    RCLCPP_WARN(
+      logger_,
+      "Exceeded time allowance before reaching the BackUp goal - Exiting BackUp");
+    return Status::FAILED;
+  }
+
+  geometry_msgs::msg::PoseStamped current_pose;
+  if (!nav2_util::getCurrentPose(
+      current_pose, *tf_, global_frame_, robot_base_frame_,
+      transform_tolerance_))
+  {
+    RCLCPP_ERROR(logger_, "Current robot pose is not available.");
+    return Status::FAILED;
+  }
+
+  // user states that teleop was successful
+  if (joy_.buttons.size() > 0 && joy_.buttons[0] > 0) {
+    stopRobot();
+    return Status::SUCCEEDED;
+  }
+
+  // user states that teleop failed
+  if (joy_.buttons.size() > 1 && joy_.buttons[1] > 0) {
+    stopRobot();
+    return Status::FAILED;
+  }
+
+  geometry_msgs::msg::Twist cmd_vel;
+  cmd_vel = computeVelocity(input_twist_);
+  vel_pub_->publish(std::move(cmd_vel));
+
+  return Status::RUNNING;
+}
+
+geometry_msgs::msg::Twist AssistedTeleop::computeVelocity(geometry_msgs::msg::Twist & twist)
```

Comment:
Merge this function into `onCycleUpdate`, after you remove the `joy` and the extra `getCurrentPose` calls, `onCycleUpdate` becomes incredibly short to basically just calling this function.